packages/backend/src/core/balances/BalanceProvider.ts

```diff
@@ -0,0 +1,57 @@
+import {
+  AssetId,
+  ChainId,
+  EthereumAddress,
+  UnixTime,
+} from '@l2beat/shared-pure'
+
+import { BalanceRecord } from '../../peripherals/database/BalanceRepository'
+import { BalanceCall } from '../../peripherals/ethereum/calls/BalanceCall'
+import { MulticallClient } from '../../peripherals/ethereum/MulticallClient'
+
+export interface HeldAsset {
+  holder: EthereumAddress
+  assetId: AssetId
+}
+
+export interface BalanceProvider {
+  fetchBalances: (
+    missingAssets: HeldAsset[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ) => Promise<BalanceRecord[]>
+  getChainId: () => ChainId
+}
+
+export class EthereumBalanceProvider implements BalanceProvider {
+  constructor(
+    private readonly multiCallClient: MulticallClient,
+    private readonly chainId: ChainId,
+  ) {}
```

Comment:
It's really hard to read without a space here

Reply:
https://github.com/l2beat/l2beat/pull/1630/commits/5a33402f11cc030022fba9b95a0d4e3a450e0f6b